stubs/WebOb/webob/acceptparse.pyi

```diff
@@ -0,0 +1,503 @@
+    @overload
+    def __add__(self, other: AcceptNoHeader | AcceptInvalidHeader) -> AcceptNoHeader: ...
+    @overload
+    def __add__(self, other: SupportsItems[str, float | tuple[float, str]]) -> AcceptValidHeader | AcceptNoHeader: ...
+    @overload
+    def __add__(self, other: _ListOrTuple[str | tuple[str, float, str] | list[Any]]) -> AcceptValidHeader | AcceptNoHeader: ...
+    @overload
+    def __radd__(self, other: None) -> AcceptNoHeader: ...
+    @overload
+    def __radd__(self, other: str) -> AcceptValidHeader | AcceptNoHeader: ...
+    @overload
+    def __radd__(self, other: AcceptValidHeader) -> AcceptValidHeader: ...
+    @overload
+    def __radd__(self, other: AcceptNoHeader | AcceptInvalidHeader) -> AcceptNoHeader: ...
+    @overload
+    def __radd__(self, other: SupportsItems[str, float | tuple[float, str]]) -> AcceptValidHeader | AcceptNoHeader: ...
+    @overload
+    def __radd__(self, other: _ListOrTuple[str | tuple[str, float, str] | list[Any]]) -> AcceptValidHeader | AcceptNoHeader: ...
+    def __bool__(self) -> Literal[False]: ...
+    def __contains__(self, offer: str) -> Literal[True]: ...
+    def __iter__(self) -> Iterator[str]: ...
+    def accept_html(self) -> bool: ...
+    @property
+    def accepts_html(self) -> bool: ...
+    def acceptable_offers(self, offers: Iterable[str | AcceptOffer]) -> list[tuple[str, float]]: ...
+    @overload
+    def best_match(self, offers: Iterable[str | tuple[str, float] | list[Any]], default_match: None = None) -> str | None: ...
+    @overload
+    def best_match(self, offers: Iterable[str | tuple[str, float] | list[Any]], default_match: str) -> str: ...
+    def quality(self, offer: str) -> float: ...
+
+class AcceptNoHeader(_AcceptInvalidOrNoHeader):
+    @property
+    def header_value(self) -> None: ...
+    def __init__(self) -> None: ...
+
+class AcceptInvalidHeader(_AcceptInvalidOrNoHeader):
+    @property
+    def header_value(self) -> str: ...
+    def __init__(self, header_value: str) -> None: ...
+
+@overload
+def create_accept_header(header_value: None) -> AcceptNoHeader: ...
+@overload
+def create_accept_header(header_value: str) -> AcceptValidHeader | AcceptInvalidHeader: ...
+
+class _AcceptProperty:
+    @overload
+    def __get__(self, __obj: None, __type: type | None = ...) -> property: ...
+    @overload
+    def __get__(self, __obj: Any, __type: type | None = ...) -> AcceptNoHeader | AcceptValidHeader | AcceptInvalidHeader: ...
+    @overload
+    def __set__(self, __obj: Any, __value: str | None) -> None: ...
+    @overload
+    def __set__(self, __obj: Any, __value: AcceptNoHeader | AcceptValidHeader | AcceptInvalidHeader) -> None: ...
+    @overload
+    def __set__(self, __obj: Any, __value: SupportsItems[str, float | tuple[float, str]]) -> None: ...
+    @overload
+    def __set__(self, __obj: Any, __value: _ListOrTuple[str | tuple[str, float, str] | list[Any]]) -> None: ...
```

Comment:
Doesn't seem like this needs overloads, you can just use a union as the argument type for `__value`.

Reply:
You are completely correct, however in this case I think a bunch of overloads reads nicer than a very large type union. I've seen overloads used like this in other places in typeshed and liked how it looks compared to a very large type union where after maybe the fourth or fifth type in the union you just stop reading, especially once the types get as complicated as the last two overloads.

So here it's separated neatly into the plain string version of the header, then the specialized objects for representing the header from WebOb, then the Mapping like way to represent this data, and then the sequence like way to represent in.

You could make a case for combining the first two overloads, but I think the other two should stay distinct. I've also debated whether or not to add the final overload at all, because it isn't completely type safe and kind of confusing to read, it just happens to be one of the supported and documented ways to pass this data in.